lib/Runtime/Library/SharedArrayBuffer.cpp

```diff
@@ -0,0 +1,577 @@
+        if (args.Info.Count < 2)
+        {
+            newLen = currentLen;
+        }
+        else
+        {
+            start = JavascriptArray::GetIndexFromVar(args[1], currentLen, scriptContext);
+
+            // If no end argument, use length as the end
+            if (args.Info.Count < 3 || args[2] == library->GetUndefined())
+            {
+                end = currentLen;
+            }
+            else
+            {
+                end = JavascriptArray::GetIndexFromVar(args[2], currentLen, scriptContext);
+            }
+
+            newLen = end > start ? end - start : 0;
+        }
+
+        // We can't have allocated an SharedArrayBuffer with byteLength > MaxArrayBufferLength.
+        // start and end are clamped to valid indices, so the new length also cannot exceed MaxArrayBufferLength.
+        // Therefore, should be safe to cast down newLen to uint32.
+        Assert(newLen < MaxSharedArrayBufferLength);
+        uint32 newbyteLength = static_cast<uint32>(newLen);
+
+        SharedArrayBuffer* newBuffer = nullptr;
+
+        if (scriptContext->GetConfig()->IsES6SpeciesEnabled())
+        {
+            Var constructorVar = JavascriptOperators::SpeciesConstructor(currentBuffer, scriptContext->GetLibrary()->GetSharedArrayBufferConstructor(), scriptContext);
+
+            JavascriptFunction* constructor = JavascriptFunction::FromVar(constructorVar);
+
+            Js::Var constructorArgs[] = { constructor, JavascriptNumber::ToVar(newbyteLength, scriptContext) };
+            Js::CallInfo constructorCallInfo(Js::CallFlags_New, _countof(constructorArgs));
+            Js::Var newVar = JavascriptOperators::NewScObject(constructor, Js::Arguments(constructorCallInfo, constructorArgs), scriptContext);
+
+            if (!SharedArrayBuffer::Is(newVar))
+            {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedSharedArrayBufferObject);
+            }
+
+            newBuffer = SharedArrayBuffer::FromVar(newVar);
+
+            if (newBuffer == currentBuffer)
+            {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedSharedArrayBufferObject);
+            }
+
+            if (newBuffer->GetByteLength() < newbyteLength)
+            {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_ArgumentOutOfRange, _u("SharedArrayBuffer.prototype.slice"));
+            }
+        }
+        else
+        {
+            newBuffer = library->CreateSharedArrayBuffer(newbyteLength);
+        }
+
+        Assert(newBuffer);
+        Assert(newBuffer->GetByteLength() >= newbyteLength);
+
+        // Don't bother doing memcpy if we aren't copying any elements
+        if (newbyteLength > 0)
+        {
+            AssertMsg(currentBuffer->GetBuffer() != nullptr, "buffer must not be null when we copy from it");
+
+            js_memcpy_s(newBuffer->GetBuffer(), newbyteLength, currentBuffer->GetBuffer() + start, newbyteLength);
```

Comment:
The SharedArrayBuffer does not get detached.

---
In reply to: [77416686](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77416686) [](ancestors = 77416686)